### ISSUE TITLE
"kubectl get -A all" hangs when using kubectl version 1.17.0

### ISSUE THREAD

**Issue by antoninbas** [kubernetes/kubernetes#86462]
I just created a new cluster using version 1.17.0 of the kubeadm & kubelet Debian packages (downloaded from https://packages.cloud.google.com/apt), and observed a strange behavior when using `kubectl` version 1.17.0.

**What happened**:
When running `kubectl get -A all` with `kubectl` version 1.17.0, the command hangs and takes about 1 minute to complete.
If instead I run `kubectl get -A pods`, the command returns immediately as expected.
I I use `kubectl` version 1.16.4 instead, the command `kubectl get -A all` also returns immediately.

**What you expected to happen**:
I would expect `kubectl get -A all` to complete in a reasonable amount of time regardless of which version of `kubectl` is used.

**How to reproduce it (as minimally and precisely as possible)**:
Create a cluster with version 1.17.0 of kubeadm and running kubelet 1.17.0 on every Node. Then do the following:
```
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.16.4/bin/linux/amd64/kubectl && chmod +x kubectl
./kubectl get -A all # returns right away
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.17.0/bin/linux/amd64/kubectl && chmod +x kubectl
./kubectl get -A all # hangs for a long time, but eventually returns successfully
./kubectl get -A pods # returns right away
```

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:20:10Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:12:17Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
I'm running Kubernetes on VM Nodes provisioned with Vagrant & Ansible.
- OS (e.g: `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="18.04.3 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.3 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
- Kernel (e.g. `uname -a`):
```
Linux k8s-node-master 4.15.0-72-generic #81-Ubuntu SMP Tue Nov 26 12:20:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

Here is a copy-paste of my terminal output (notice the `kubectl` invocation that takes a minute to complete):
```
vagrant@k8s-node-master:~$ curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.16.4/bin/linux/amd64/kubectl && chmod +x kubectl
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 44.5M  100 44.5M    0     0  35.5M      0  0:00:01  0:00:01 --:--:-- 35.5M
vagrant@k8s-node-master:~$ time ./kubectl get -A all -o name
pod/coredns-6955765f44-m4qjj
pod/coredns-6955765f44-r74kd
pod/etcd-k8s-node-master
pod/kube-apiserver-k8s-node-master
pod/kube-controller-manager-k8s-node-master
pod/kube-proxy-7r47f
pod/kube-proxy-mcz7t
pod/kube-scheduler-k8s-node-master
service/kubernetes
service/kube-dns
daemonset.apps/kube-proxy
deployment.apps/coredns
replicaset.apps/coredns-6955765f44

real	0m0.219s
user	0m0.137s
sys	0m0.033s
vagrant@k8s-node-master:~$ curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.17.0/bin/linux/amd64/kubectl && chmod +x kubectl
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 41.4M  100 41.4M    0     0  35.3M      0  0:00:01  0:00:01 --:--:-- 35.3M
vagrant@k8s-node-master:~$ time ./kubectl get -A all -o name
pod/coredns-6955765f44-m4qjj
pod/coredns-6955765f44-r74kd
pod/etcd-k8s-node-master
pod/kube-apiserver-k8s-node-master
pod/kube-controller-manager-k8s-node-master
pod/kube-proxy-7r47f
pod/kube-proxy-mcz7t
pod/kube-scheduler-k8s-node-master
service/kubernetes
service/kube-dns
daemonset.apps/kube-proxy
deployment.apps/coredns
replicaset.apps/coredns-6955765f44

real	0m57.104s
user	0m0.277s
sys	0m0.115s
vagrant@k8s-node-master:~$ time ./kubectl get -A pods -o name
pod/coredns-6955765f44-m4qjj
pod/coredns-6955765f44-r74kd
pod/etcd-k8s-node-master
pod/kube-apiserver-k8s-node-master
pod/kube-controller-manager-k8s-node-master
pod/kube-proxy-7r47f
pod/kube-proxy-mcz7t
pod/kube-scheduler-k8s-node-master

real	0m0.105s
user	0m0.060s
sys	0m0.035s
```

**Comment by liggitt** [comment]
What does trace logging show is taking a long time (add `--v=6`)

**Comment by antoninbas** [comment]
@liggitt Thanks for the quick follow-up. A lot of messages like this one:
```
I1220 03:33:02.306615   22091 cached_discovery.go:87] failed to write cache to /home/vagrant/.kube/cache/discovery/192.168.77.100_6443/node.k8s.io/v1beta1/serverresources.json due to mkdir /home/vagrant/.kube/cache: permission denied
```
which makes sense since the `.kube` folder was created by the root user in my Ansible playbook and is owned by root. So related to this older issue: https://github.com/kubernetes/kubernetes/issues/59356. Funny how I don't have this issue with older kubectl versions.

**Comment by liggitt** [comment]
Older versions of kubectl do the same thing... that behavior hasn't changed in years. Working without a cache means most calls fully rediscover available APIs, which unsurprisingly has an impact on performance.

**Comment by antoninbas** [comment]
Yes you're right I see the same messages with kubectl 1.16.4, but for some reason the request still completes much faster. I see these messages with kubectl 1.17.0:
```
I1220 03:43:13.144061   24156 request.go:565] Throttling request took 2.397891481s, request: GET:https://192.168.77.100:6443/apis/autoscaling/v1?timeout=32s
I1220 03:43:13.145566   24156 round_trippers.go:443] GET https://192.168.77.100:6443/apis/autoscaling/v1?timeout=32s 200 OK in 1 milliseconds
I1220 03:43:13.145858   24156 cached_discovery.go:87] failed to write cache to /home/vagrant/.kube/cache/discovery/192.168.77.100_6443/autoscaling/v1/serverresources.json due to mkdir /home/vagrant/.kube/cache: permission denied
I1220 03:43:13.339375   24156 request.go:565] Throttling request took 2.593197733s, request: GET:https://192.168.77.100:6443/apis/autoscaling/v2beta1?timeout=32s
I1220 03:43:13.340391   24156 round_trippers.go:443] GET https://192.168.77.100:6443/apis/autoscaling/v2beta1?timeout=32s 200 OK in 0 milliseconds
I1220 03:43:13.340605   24156 cached_discovery.go:87] failed to write cache to /home/vagrant/.kube/cache/discovery/192.168.77.100_6443/autoscaling/v2beta1/serverresources.json due to mkdir /home/vagrant/.kube/cache: permission denied
I1220 03:43:13.539727   24156 request.go:565] Throttling request took 2.793536036s, request: GET:https://192.168.77.100:6443/apis/autoscaling/v2beta2?timeout=32s
I1220 03:43:13.540953   24156 round_trippers.go:443] GET https://192.168.77.100:6443/apis/autoscaling/v2beta2?timeout=32s 200 OK in 1 milliseconds
I1220 03:43:13.541222   24156 cached_discovery.go:87] failed to write cache to /home/vagrant/.kube/cache/discovery/192.168.77.100_6443/autoscaling/v2beta2/serverresources.json due to mkdir /home/vagrant/.kube/cache: permission denied
I1220 03:43:13.739445   24156 request.go:565] Throttling request took 2.993227596s, request: GET:https://192.168.77.100:6443/apis/storage.k8s.io/v1beta1?timeout=32s
I1220 03:43:13.740712   24156 round_trippers.go:443] GET https://192.168.77.100:6443/apis/storage.k8s.io/v1beta1?timeout=32s 200 OK in 1 milliseconds
I1220 03:43:13.741400   24156 cached_discovery.go:87] failed to write cache to /home/vagrant/.kube/cache/discovery/192.168.77.100_6443/storage.k8s.io/v1beta1/serverresources.json due to mkdir /home/vagrant/.kube/cache: permission denied
I1220 03:43:13.942011   24156 request.go:565] Throttling request took 3.195718695s, request: GET:https://192.168.77.100:6443/apis/admissionregistration.k8s.io/v1?timeout=32s
```

Some throttling happening causing the command to take much longer to execute?

**Comment by liggitt** [comment]
ah, that would be https://github.com/kubernetes/kubernetes/pull/86168, which is already merged into the 1.17 branch in https://github.com/kubernetes/kubernetes/pull/86170 and will be fixed in 1.17.1

**Comment by antoninbas** [comment]
Sorry for the duplicate issue and thanks for the pointer! I will "fix" the ownership of the ".kube" directory to avoid the issue until 1.17.1 is released.
